src/ATen/native/xpu/sycl/SortingKernels.h

```diff
@@ -403,6 +403,124 @@ void segmented_radix_sort_pairs_kernel(
   }
 }
 
+// ======================= group radix select =======================
+
+template <int n, typename index_t>
+inline index_t make_alignment_n(index_t size) {
+  return (size + n - 1) / n * n;
+}
+
+template <typename method_t, typename key_t, typename value_t>
+struct SegmentedGroupRadixSelectPairsFunctor
+    : public __SYCL_KER_CONFIG_CONVENTION__ {
+  enum {
+    MAX_KV_BYTES = std::max(sizeof(key_t), sizeof(value_t)),
+  };
+
+  [[intel::reqd_sub_group_size(method_t::SUBGROUP_SIZE)]] void operator()(
+      sycl::nd_item<1> item) const {
+    int seg_idx = item.get_group(0);
+    int seg_offset = seg_idx * nelements_;
+    auto method = method_t(item, slm_);
+
+    auto keys_in_seg = keys_in_ + seg_offset;
+    auto values_in_seg =
+        values_in_ == nullptr ? nullptr : values_in_ + seg_offset;
+
+    key_t* keys_temp = reinterpret_cast<key_t*>(
+        slm_.template get_multi_ptr<sycl::access::decorated::no>().get() +
+        make_alignment_n<MAX_KV_BYTES>(method_t::LocalMemorySize()));
+    value_t* values_temp = reinterpret_cast<value_t*>(
+        reinterpret_cast<char*>(keys_temp) +
+        make_alignment_n<MAX_KV_BYTES>(k_ * sizeof(key_t)));
+
+    method.load_keys(keys_in_seg, nelements_);
+    method.load_values(values_in_seg, nelements_);
+
+    int num_start = method_t::PROCESSING_LENGTH;
+    while (num_start < nelements_) {
+      method.topk(KeyTraits<key_t>::endbit(), 0, k_, keys_temp, values_temp);
+      item.barrier(sycl_local_fence);
+      method.topk_append_keys(
+          keys_in_seg, keys_temp, nelements_, num_start, k_);
+      method.topk_append_values(
+          values_in_seg, values_temp, nelements_, num_start, k_);
+      num_start += method_t::PROCESSING_LENGTH - k_;
+      item.barrier(sycl_local_fence);
+    }
+
+    method.topk(
+        KeyTraits<key_t>::endbit(),
+        0,
+        k_,
+        keys_out_ + seg_idx * k_,
+        values_out_ + seg_idx * k_);
+  }
+
+  void sycl_ker_config_convention(sycl::handler& cgh) {
+    slm_ = sycl_local_acc_t<char>(
+        make_alignment_n<MAX_KV_BYTES>(method_t::LocalMemorySize()) +
+            make_alignment_n<MAX_KV_BYTES>(k_ * sizeof(key_t)) +
+            k_ * sizeof(value_t),
```

Comment:
Do we need to check the size, which should be less than hardware limitation?

Reply:
No need, LocalMemorySize returns a fixed small value, k_ is limited within the work group size.